Add GSN intro guides (#10)
OpenZeppelin/openzeppelin-docs: modules/ROOT/pages/gsn/what-is-the-gsn.adoc

= What is the Gas Station Network?

> *https://gsn.openzeppelin.com[Website]*

Reviewer: The link by itself looks sort of weird on the docsite, perhaps we could put it on the title?

User onboarding is one of the hottest topics in Ethereum. UI/UX along with scalability have been identified as the main problems that prevent adoption. Meta-transactions are a key component to improve the user experience.

> Here you'll learn about what the Gas Station Network (GSN) is and how it solves this problem, by allowing users of a contract to send transactions without needing Ether to pay for gas.

Reviewer: ```suggestion
Here you'll learn what the Gas Station Network (GSN) is and how it solves this problem, by allowing users of a contract to send transactions without needing Ether to pay for gas.
```

== Sending gas-less transactions

> All Ethereum transactions use gas, and the sender of each transaction must have enough Ether to pay for the gas spent. Even though these gas costs are low for basic transactions (a couple of cents), getting this Ether is no easy task: dApp users often need to go through Know Your Customer and Anti Money-Laundering processes (KYC & AML), which not only takes time but often involves sending a selfie holding their passport over the Internet (!).

Reviewer: ```suggestion
All Ethereum transactions use gas, and the sender of each transaction must have enough Ether to pay for the gas spent. Even though these gas costs are low for basic transactions (a couple of cents), getting Ether is no easy task: dApp users often need to go through Know Your Customer and Anti Money-Laundering processes (KYC & AML), which not only takes time but often involves sending a selfie holding their passport over the Internet (!).
```

One of RelayHub's jobs is to act as a, well, _hub_ for all relayers: they will advertise their services on this contract, and your users will query it to find the relayer that best suits their purposes. This is out of scope for this guide however, and is not something you need to worry about when writing a recipient contract. If you want to learn more about sending transactions via relayers, head to our https://github.com/OpenZeppelin/openzeppelin-gsn-provider[GSNProvider guide].

The other key task RelayHub carries out is the actual _relaying of transactions_, the sole purpose behind this whole system. Instead of calling a function in your contract directly, your users will request a relayer to do it for them, who will then execute RelayHub's relayCall function. RelayHub will verify that the transaction is legitimate (protecting both users and recipients from dishonest relayers), and then call into your contract as originally requested by your user. This requires your recipient trusting RelayHub to do the right thing, but since it is a smart contract, this is as simple as reading its source code!

> NOTE: The RelayHub address will be the same in each network. Right now, the latest relay hub is live with this address `0xD216153c06E857cD7f72665E0aF1d7D82172F494`.

Reviewer: ```suggestion
NOTE: The RelayHub address will be the same in each network. Right now, the latest relay hub is live at this address: https://etherscan.io/address/0xD216153c06E857cD7f72665E0aF1d7D82172F494[`0xD216153c06E857cD7f72665E0aF1d7D82172F494`].
```

== Receiving a relayed call

We've mentioned how the RelayHub, and not your user, is the one that actually ends up calling a function in your contract. We will refer to this as the _relayed call_. Your contract needs to be set up to accept relayed calls from the hub. In particular, it needs to be able to answer whether it will pay for a given relayed call, and run some bookeeping to make sure a malicious user cannot abuse it. It also needs to unwrap a relayed call in order to process it.

The OpenZeppelin Contracts package includes a number of utilities to make receiving relayed calls as easy as developing a regular Solidity contract, without needing to worry about the low level details. It also ships with two built-in strategies for managing relayed call subsidies.

== Payment

By now you may be wondering how exactly relayers charge their recipients for gas costs and service fees. The answer is simple: each recipient must have funds deposited on RelayHub in advance, and payment is automatically handled on each relayed call.

> You can head to the https://gsn.openzeppelin.com/recipients[GSN Recipient Tool] to check and top-up your contracts' balance, view previous charges, or do all of this programatically by calling `IRelayHub.depositFor` and `IRelayHub.balanceOf`.

Reviewer: Should we mention this can also be done via the `oz-gsn` cli?

== Summing up

To recap, there are four main actors in the GSN:

- A singleton *RelayHub* contract, that coordinates the entire network. All relayed transactions need to go through it, and all relayer stakes and recipients funds are held by it.
- A decentralized set of *Relayers*. A relayer must register itself on the RelayHub by staking funds, as the hub penalizes it if it misbehaves.
- A set of *Recipients*, which are the dapp contracts that pay for their users' transactions. Recipients need to deposit funds on the hub in order to pay for the relayed calls, and need to adhere to a specific interface to accept the meta transactions.
- The end *Users* of the recipient contracts. Instead of sending regular Ethereum transactions, these users can just sign a transaction with any key, and send it off-chain to a Relayer to be sent to the network.

== Further reading

* The https://medium.com/@rrecuero/eth-onboarding-solution-90607fb81380[GSN announcement post] provides a good *overview of the system*, along with some use cases to take inspiration from.

> * If you want to learn how to use *OpenZeppelin Contract's pre-made accept and charge strategies*, we'll add an advanced GSN recipients guide soon.

Reviewer: ```suggestion
* To learn how to use OpenZeppelin Contracts to *build a GSN-capable contract*, head to the https://docs.openzeppelin.com/contracts/2.x/gsn[GSN basics guide].
* If you want to learn how to use OpenZeppelin Contracts' *pre-made accept and charge strategies*, we'll add an advanced GSN recipients guide soon.
```
